### PR TITLE
feat: Add API for git docs and supported shh keys

### LINKED ISSUE #14563: [Task]: Modular git APIs in BE
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

As of today (2022.06.15), `ssh-keypair` has multiple responsibilities:

- Fetching git-app associated keys
- Fetching available keyTypes (protocols)

In some scenarios:

- connect: after loading the modal, keyTypes is not available, so a network call has to be made to `ssh-keypair.`
- import: cannot make a `ssh-keypair` call because application Id doesn’t exist yet, so API fails.

### Before the key has been generated:

```json
{
    "responseMeta": {
        "status": 400,
        "success": false,
        "error": {
            "code": 4031,
            "message": "Git configuration is invalid. Details: Can't find valid SSH key. Please configure the application with git",
            "errorType": "GIT_CONFIGURATION_ERROR"
        }
    }
}
```

But a PR has been created that will return:

```json
{
    "responseMeta": {
        "status": 200,
        "success": true
    },
    "data": {
        "publicKey": "",
        "docUrl": "https://docs.appsmith.com/core-concepts/version-control-with-git/connecting-to-git-repository#generating-a-deploy-key",
        "gitSupportedSSHKeyType": [
            {
                "protocolName": "ECDSA",
                "platFormSupported": "",
                "keySize": 256
            },
            {
                "protocolName": "RSA",
                "platFormSupported": "Azure Devops",
                "keySize": 4096
            }
        ]
    }
}
```

[https://github.com/appsmithorg/appsmith/pull/14340](https://github.com/appsmithorg/appsmith/pull/14340)

### After the key has been generated:

```json
{
    "responseMeta": {
        "status": 201,
        "success": true
    },
    "data": {
        "publicKey": "ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBGp0NVasvGP8+FCe1VPy+z5L0KvbDHPGHBYgYzMxT92cIrpoLmAOhlGOq78M+SByabUlz+bdjOAv3td2IAxFPi4= appsmith\n",
        "docUrl": "https://docs.appsmith.com/core-concepts/version-control-with-git/connecting-to-git-repository#generating-a-deploy-key",
        "gitSupportedSSHKeyType": [
            {
                "protocolName": "ECDSA",
                "platFormSupported": "",
                "keySize": 256
            },
            {
                "protocolName": "RSA",
                "platFormSupported": "Azure Devops",
                "keySize": 4096
            }
        ]
    }
}
```

### No application id exist yet:

`/api/v1/git/import/keys`

```json
{
    "responseMeta": {
        "status": 200,
        "success": true
    },
    "data": {
        "publicKey": "ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBA7QfklitvVKuQsihe5+HIr2p27wp+xOOPLPOC/9HROXJJKat2EATUJVFo3wpyoXd4NT00kPq+I9kLDR5CZy9S4= appsmith\n",
        "docUrl": "https://docs.appsmith.com/core-concepts/version-control-with-git/connecting-to-git-repository#generating-a-deploy-key",
        "regeneratedKey": false
    }
}
```

- This creates an issue for azure import, since there is no RSA type to be regenerated.
- The call is GET; should be POST as we are creating a new resource.

On introducing `ssh-keypair` on import modal load, it fails because application Id doesn’t exist.

`/api/v1/applications/ssh-keypair/`

```json
{
    "responseMeta": {
        "status": 404,
        "success": false,
        "error": {
            "code": 4027,
            "message": "Unable to find application ssh-keypair",
            "errorType": "INTERNAL_ERROR"
        }
    }
}
```

## Requirements

- `/api/v1/git/doc-urls` for all the urls that belong to git.
- `/api/v1/git/key-types` for all of the key types that ssh key regeneration support

### PR DESCRIPTION
**Pull request by AnaghHegde**
## Description

Present, GET /api/v1/git/ssh-keypair has multiple responsibilities:
Fetching git-app associated keys
Fetching available keyTypes (protocols)

In some scenarios:
connect: after loading the modal, keyTypes is not available, so a network call has to be made to ssh-keypair.
import: cannot make a ssh-keypair call because application Id doesn’t exist yet, so API fails.

For the above reasons, we are adding two new APIs for git.

Fixes #14563 

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Locally
- JUnit Test - pending

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/constants/ErrorReferenceDocUrl.java [hunk 1]
```diff
@@ -3,10 +3,20 @@
 import lombok.Getter;
 
 @Getter
```
Comment (abhvsn) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/external/constants/ErrorReferenceDocUrl.java:5]: Is `@Getter` required anymore?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/ce/GitControllerCE.java [hunk 2]
```diff
@@ -227,11 +230,17 @@ public Mono<ResponseDTO<Application>> discardChanges(@PathVariable String defaul
                 .map(result -> new ResponseDTO<>((HttpStatus.OK.value()), result, null));
     }
 
-    @GetMapping("/protocol/keys")
+    @GetMapping("/protocol/key-types")
```
Comment (abhvsn) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/ce/GitControllerCE.java:233]: Is this used in client-side code base anywhere today? If yes we might need to support previous path for backward compatibility.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/GitServiceCEImpl.java [hunk 3]
```diff
@@ -2171,6 +2173,19 @@ public Mono<Application> discardChanges(String defaultApplicationId, String bran
         );
     }
 
+    @Override
```
Comment (abhvsn) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/GitServiceCEImpl.java:2176]: Let's add JavaDoc comments for clarity on why we wanted this API in first place.

**Review by abhvsn — COMMENTED** [review]
Very minor nits, overall LGTM

File: app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/ce/GitControllerCE.java (hunk 2 above)
Comment (AnaghHegde) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/ce/GitControllerCE.java:233]: This is not used anywhere in FE at the movement.

**Review by abhvsn — APPROVED** [review]

**Pull request merged**